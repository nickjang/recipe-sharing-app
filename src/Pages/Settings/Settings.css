.settings output {
  display: block;
  margin-bottom: 1rem;
}

.settings input {
  border: 1px solid #d3d3d3;
}
