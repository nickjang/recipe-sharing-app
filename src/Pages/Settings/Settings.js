import React, { Component } from 'react';
import AccountInput from '../../Components/AccountInput/AccountInput';
import AuthApiService from '../../services/auth-api-service';
import './Settings.css';

class Settings extends Component {
  state = {
    email: {
      value: '',
      touched: false,
      updated: false
    },
    password: {
      value: '',
      touched: false,
      updated: false
    },
    newEmail: {
      value: '',
      touched: false,
    },
    newPassword: {
      value: '',
      touched: false
    },
    error: ''
  }

  emailRef = React.createRef();

  updateEmail = (email) => {
    this.setState({ email: { value: email, touched: true, updated: this.state.email.updated } });
  }

  updatePassword = (password) => {
    this.setState({ password: { value: password, touched: true } });
  }

  updateNewEmail = (newEmail) => {
    this.setState({ newEmail: { value: newEmail, touched: true } });
  }

  updateNewPassword = (newPassword) => {
    this.setState({ newPassword: { value: newPassword, touched: true } });
  }

  validateEmail = (isCurrEmail) => {
    let email;
    if (isCurrEmail)
      email = this.state.email.value.trim();
    else
      email = this.state.newEmail.value.trim();

    if (!email) return 'Email cannot be empty.';
    if (!email.includes('@')) return 'Email must include \'@\'.';
    if (!email.split('@')[1].includes('.')) return 'Email must include a \'.\'.';
    if (email.length < 5) return 'Please enter a valid email.';
    return;
  }

  validatePassword = (isCurrPassword) => {
    let password;
    if (isCurrPassword)
      password = this.state.password.value.trim();
    else
      password = this.state.newPassword.value.trim();

    if (!password) return 'Password cannot be empty.';
    if (password.length < 8) return 'Password must be at least 8 characters.';
    if (!(/\d/.test(password))) return 'Password must contain at least one number.';
    return;
  }

  handleSubmit = (e) => {
    e.preventDefault();

    if (this.state.email.value === 'demo@demo.com')
      return this.setState({ error: 'Cannot update demo account' });

    this.setState(
      {
        error: '',
        email: {
          ...this.state.email,
          updated: false
        }
      },
      () => {
        AuthApiService.updateUser(
          this.state.newEmail.value,
          this.state.newPassword.value
        )
          .then(user => {
            this.setState(
              {
                email: {
                  value: user.email,
                  touched: this.state.email.touched,
                  updated: user.email !== this.state.email.value
                },
                password: {
                  ...this.state.password,
                  updated: !this.validatePassword(this.state.newPassword)
                }
              }
            );
          })
          .catch(res => {
            this.setState({ error: res.error })
          });
      }
    );
  }

  handleDeleteAccount = (e) => {
    e.preventDefault();

    if (this.state.email.value === 'demo@demo.com')
      return this.setState({ error: 'Cannot delete demo account' });

    this.setState(
      { error: '' },
      () => {
        AuthApiService.deleteUser()
          .catch(res => {
            this.setState({ error: res.error })
          });
      }
    );
  }

  componentDidMount() {
    if (this.emailRef.current) this.emailRef.current.focus();
  }

  render() {
    const updateMessage = (
      `${this.state.email.updated ? `Updated email to ${this.state.email.value}.` : ''}
      ${this.state.password.updated ? 'Updated password.' : ''}`
    );
    return (
      <article className='form settings'>
        <h2 className='rs-title'>Settings</h2>
        <section>
          <h3 className='rs-title'>Account Settings</h3>
          <output
            form='account-settings-form'
            className={`form-status ${this.state.error ? 'fail-status' : ''}`}
          > {this.state.error || updateMessage}
          </output>
          <form id='account-settings-form'>
            <AccountInput
              form='account-settings-form'
              id='email'
              type='email'
              touched={this.state.email.touched}
              ref={el => this.emailRef = el}
              validate={() => this.validateEmail(true)}
              update={this.updateEmail}
              hint={<sup>*</sup>} />
            <AccountInput
              form='account-settings-form'
              id='password'
              type='password'
              touched={this.state.password.touched}
              validate={() => this.validatePassword(true)}
              update={this.updatePassword}
              hint={<sup>*</sup>} />
            <AccountInput
              form='account-settings-form'
              id='new-email'
              type='email'
              touched={this.state.newEmail.touched}
              validate={() => this.validateEmail(false)}
              update={this.updateNewEmail} />
            <AccountInput
              form='account-settings-form'
              id='new-password'
              type='password'
              touched={this.state.newPassword.touched}
              validate={() => this.validatePassword(false)}
              update={this.updateNewPassword} />
            <button
              className='rs-btn rs-btn-success mt-1 mr-1'
              type='submit'
              form='account-settings-form'
              onClick={(e) => { this.handleSubmit(e) }}
              disabled={this.validateEmail(true) ||
                this.validatePassword(true) ||
                (this.validateEmail(false) && this.validatePassword(false))}
            > Submit
            </button>
            <button
              className='rs-btn rs-btn-reset'
              type='reset'
              form='account-settings-form'
              onClick={(e) => { this.handleDeleteAccount(e) }}
              disabled={this.validateEmail(true) || this.validatePassword(true)}
            > Delete Account
            </button>
          </form>
        </section>
      </article>
    );
  }
}

export default Settings;