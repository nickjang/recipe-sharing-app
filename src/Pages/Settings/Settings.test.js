import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Settings from './Settings';

describe(`Settings component`, () => {
  it('renders Settings by default', () => {
    const wrapper = shallow(<Settings />);
    expect(toJson(wrapper)).toMatchSnapshot()
  });
})