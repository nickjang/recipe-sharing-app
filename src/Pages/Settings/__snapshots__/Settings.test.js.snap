// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings component renders Settings by default 1`] = `
<article
  className="form settings"
>
  <h2
    className="rs-title"
  >
    Settings
  </h2>
  <section>
    <h3
      className="rs-title"
    >
      Account Settings
    </h3>
    <output
      className="form-status "
      form="account-settings-form"
    >
       
      
      
    </output>
    <form
      id="account-settings-form"
    >
      <AccountInput
        form="account-settings-form"
        hint={
          <sup>
            *
          </sup>
        }
        id="email"
        inputRef={null}
        touched={false}
        type="email"
        update={[Function]}
        validate={[Function]}
      />
      <AccountInput
        form="account-settings-form"
        hint={
          <sup>
            *
          </sup>
        }
        id="password"
        inputRef={null}
        touched={false}
        type="password"
        update={[Function]}
        validate={[Function]}
      />
      <AccountInput
        form="account-settings-form"
        hint=""
        id="new-email"
        inputRef={null}
        touched={false}
        type="email"
        update={[Function]}
        validate={[Function]}
      />
      <AccountInput
        form="account-settings-form"
        hint=""
        id="new-password"
        inputRef={null}
        touched={false}
        type="password"
        update={[Function]}
        validate={[Function]}
      />
      <button
        className="rs-btn rs-btn-success mt-1 mr-1"
        disabled="Email cannot be empty."
        form="account-settings-form"
        onClick={[Function]}
        type="submit"
      >
         Submit
      </button>
      <button
        className="rs-btn rs-btn-reset"
        disabled="Email cannot be empty."
        form="account-settings-form"
        onClick={[Function]}
        type="reset"
      >
         Delete Account
      </button>
    </form>
  </section>
</article>
`;
